teamcity: failed test: TestParallel invalid retryable err

The following tests appear to have failed on master (testrace): TestParallel, TestParallel/create_stats

You may want to check [for open issues](https://github.com/cockroachdb/cockroach/issues?q=is%3Aissue+is%3Aopen+TestParallel).

[#1459621](https://teamcity.cockroachdb.com/viewLog.html?buildId=1459621):


```
TestParallel/create_stats
...torage/stores.go:181 +0x1e8
github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal.func1(0x5cad860, 0xc00631abd0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:916 +0x254
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr(0xc0014e65a0, 0x5cad860, 0xc00631abd0, 0x4e21843, 0x10, 0xc008d21438, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:300 +0xf9
github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal(0xc000530500, 0x5cad860, 0xc00631abd0, 0xc003582f00, 0xc00631abd0, 0x66d28c, 0xc00004e000)
	/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:904 +0x2b0
github.com/cockroachdb/cockroach/pkg/server.(*Node).Batch(0xc000530500, 0x5cad860, 0xc00631aba0, 0xc003582f00, 0xc000530500, 0xc0002acda0, 0x5c4b100)
	/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:942 +0xb0
github.com/cockroachdb/cockroach/pkg/roachpb._Internal_Batch_Handler.func1(0x5cad860, 0xc00631aba0, 0x4de4e80, 0xc003582f00, 0x4de4e80, 0xc003582f00, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/roachpb/api.pb.go:9981 +0xa2
github.com/cockroachdb/cockroach/vendor/github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1(0x5cad860, 0xc00631aba0, 0x4de4e80, 0xc003582f00, 0xc006e7ce60, 0xc006e7ce80, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc/server.go:44 +0xbc9
github.com/cockroachdb/cockroach/pkg/rpc.NewServerWithInterceptor.func2(0x5cad860, 0xc00631aba0, 0x4de4e80, 0xc003582f00, 0xc006e7ce60, 0xc006e7ce80, 0xc006e7ce80, 0x4247fe0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:221 +0x10f
github.com/cockroachdb/cockroach/pkg/rpc.NewServerWithInterceptor.func4(0x5cad860, 0xc00631aba0, 0x4de4e80, 0xc003582f00, 0xc006e7ce60, 0xc006e7ce80, 0x86f1da0, 0xc0003157c0, 0xc0070759c8, 0x670a58)
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:251 +0xea
github.com/cockroachdb/cockroach/pkg/roachpb._Internal_Batch_Handler(0x4db1720, 0xc000530500, 0x5cad860, 0xc00631aba0, 0xc005d97aa0, 0xc00037bd40, 0x5cad860, 0xc00631aba0, 0xc003652600, 0x200)
	/go/src/github.com/cockroachdb/cockroach/pkg/roachpb/api.pb.go:9983 +0x1d9
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).processUnaryRPC(0xc001172000, 0x5d0a800, 0xc0034bcd80, 0xc00174a300, 0xc0018ec2d0, 0x844b7c0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:999 +0x99c
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).handleStream(0xc001172000, 0x5d0a800, 0xc0034bcd80, 0xc00174a300, 0x0)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1279 +0x1352
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc00169d390, 0xc001172000, 0x5d0a800, 0xc0034bcd80, 0xc00174a300)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:717 +0xad
created by github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).serveStreams.func1
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:715 +0xb9


****************************************************************************

This node experienced a fatal error (printed above), and as a result the
process is terminating.

Fatal errors can occur due to faulty hardware (disks, memory, clocks) or a
problem in CockroachDB. With your help, the support team at Cockroach Labs
will try to determine the root cause, recommend next steps, and we can
improve CockroachDB based on your report.

Please submit a crash report by following the instructions here:

    https://github.com/cockroachdb/cockroach/issues/new/choose

If you would rather not post publicly, please contact us directly at:

    support@cockroachlabs.com

The Cockroach Labs team appreciates your feedback.



TestParallel
--- FAIL: testrace/TestParallel (0.000s)
Test ended in panic.





```

Please assign, take a look and update the issue accordingly.

--- a/pkg/kv/txn_interceptor_span_refresher.go
+++ b/pkg/kv/txn_interceptor_span_refresher.go
@@ -196,10 +196,31 @@ func (sr *txnSpanRefresher) sendLockedWithRefreshAttempts(
 	ctx context.Context, ba roachpb.BatchRequest, maxRefreshAttempts int,
 ) (_ *roachpb.BatchResponse, _ *roachpb.Error, largestRefreshTS hlc.Timestamp) {
 	br, pErr := sr.wrapped.SendLocked(ctx, ba)
-	if pErr != nil && maxRefreshAttempts > 0 {
-		br, pErr, largestRefreshTS = sr.maybeRetrySend(ctx, ba, br, pErr, maxRefreshAttempts)
+	if pErr != nil {
+		if unwrappedErr, ok := maybeUnwrapMixedSuccessErr(pErr); ok {
+			log.VEventf(ctx, 2, "got partial success; cannot retry %s (pErr=%s)", ba, unwrappedErr)
+			return nil, unwrappedErr, hlc.Timestamp{}
+		}
+		if maxRefreshAttempts > 0 {
+			br, pErr, largestRefreshTS = sr.maybeRetrySend(ctx, ba, br, pErr, maxRefreshAttempts)
+		}
 	}
 	return br, pErr, largestRefreshTS
+}
+
+// maybeUnwrapMixedSuccessErr attempts to unwrap a mixed success error, and
+// returns whether it was successful or not. With mixed success, we can't
+// attempt a retry without potentially succeeding at the same conditional put or
+// increment request twice; so we return the wrapped error instead. Because the
+// dist sender splits up batches to send to multiple ranges in parallel, and
+// then combines the results, partial success makes it very difficult to
+// determine what can be retried.
+func maybeUnwrapMixedSuccessErr(pErr *roachpb.Error) (*roachpb.Error, bool) {
+	if mse, ok := pErr.GetDetail().(*roachpb.MixedSuccessError); ok {
+		pErr.SetDetail(mse.GetWrapped())
+		return pErr, true
+	}
+	return pErr, false
 }
 
 // maybeRetrySend attempts to catch serializable errors and avoid them by
@@ -213,18 +234,6 @@ func (sr *txnSpanRefresher) maybeRetrySend(
 	pErr *roachpb.Error,
 	maxRefreshAttempts int,
 ) (*roachpb.BatchResponse, *roachpb.Error, hlc.Timestamp) {
-	// With mixed success, we can't attempt a retry without potentially
-	// succeeding at the same conditional put or increment request
-	// twice; return the wrapped error instead. Because the dist sender
-	// splits up batches to send to multiple ranges in parallel, and
-	// then combines the results, partial success makes it very
-	// difficult to determine what can be retried.
-	if mse, ok := pErr.GetDetail().(*roachpb.MixedSuccessError); ok {
-		pErr.SetDetail(mse.GetWrapped())
-		log.VEventf(ctx, 2, "got partial success; cannot retry %s (pErr=%s)", ba, pErr)
-		return nil, pErr, hlc.Timestamp{}
-	}
-
 	// Check for an error which can be retried after updating spans.
 	canRetryTxn, retryTxn := roachpb.CanTransactionRetryAtRefreshedTimestamp(ctx, pErr)
 	if !canRetryTxn || !sr.canAutoRetry {

--- a/pkg/kv/txn_interceptor_span_refresher.go
+++ b/pkg/kv/txn_interceptor_span_refresher.go
@@ -39,11 +39,69 @@ var MaxTxnRefreshSpansBytes = settings.RegisterIntSetting(
 	256*1000,
 )
 
-// txnSpanRefresher is a txnInterceptor that collects the read spans
-// of a serializable transaction in the event we get a serializable
-// retry error. We can use the set of read spans to avoid retrying
-// the transaction if all the spans can be updated to the current
-// transaction timestamp.
+// txnSpanRefresher is a txnInterceptor that collects the read spans of a
+// serializable transaction in the event it gets a serializable retry error. It
+// can then use the set of read spans to avoid retrying the transaction if all
+// the spans can be updated to the current transaction timestamp.
+//
+// Serializable isolation mandates that transactions appear to have occurred in
+// some total order, where none of their component sub-operations appear to have
+// interleaved with sub-operations from other transactions. CockroachDB enforces
+// this isolation level by ensuring that all of a transaction's reads and writes
+// are performed at the same HLC timestamp. This timestamp is referred to as the
+// transaction's commit timestamp.
+//
+// As a transaction in CockroachDB executes at a certain provisional commit
+// timestamp, it lays down intents at this timestamp for any write operations
+// and ratchets various timestamp cache entries to this timestamp for any read
+// operations. If a transaction performs all of its reads and writes and is able
+// to commit at its original provisional commit timestamp then it may go ahead
+// and do so. However, for a number of reasons including conflicting reads and
+// writes, a transaction may discover that its provisional commit timestamp is
+// too low and that it needs to move this timestamp forward to commit.
+//
+// This poses a problem for operations that the transaction has already
+// completed at lower timestamps. Are the effects of these operations still
+// valid? The transaction is always free to perform a full restart at a higher
+// epoch, but this often requires iterating in a client-side retry loop and
+// performing all of the transaction's operations again. Intents are maintained
+// across retries to improve the chance that later epochs succeed, but it is
+// vastly preferable to avoid re-issuing these operations. Instead, it would be
+// ideal if the transaction could "move" each of its operations to its new
+// provisional commit timestamp without redoing them entirely.
+//
+// Only a single write intent can exist on a key and no reads are allowed above
+// the intent's timestamp until the intent is resolved, so a transaction is free
+// to move any of its intent to a higher timestamp. In fact, a synchronous
+// rewrite of these intents isn't even necessary because intent resolution will
+// already rewrite the intents at higher timestamp if necessary. So, moving
+// write intents to a higher timestamp can be performed implicitly by committing
+// their transaction at a higher timestamp. However, unlike intents created by
+// writes, timestamp cache entries created by reads only prevent writes on
+// overlapping keys from being written at or below their timestamp; they do
+// nothing to prevent writes on overlapping keys from being written above their
+// timestamp. This means that a transaction is not free to blindly move its
+// reads to a higher timestamp because writes from other transaction may have
+// already invalidated them. In effect, this means that transactions acquire
+// pessimistic write locks and optimistic read locks.
+//
+// The txnSpanRefresher is in charge of detecting when a transaction may want to
+// move its provisional commit timestamp forward and determining whether doing
+// so is safe given the reads that it has performed (i.e. its "optimistic read
+// locks"). When the interceptor decides to attempt to move a transaction's
+// timestamp forward, it first "refreshes" each of its reads. This refreshing
+// step revisits all of the key spans that the transaction has read and checks
+// whether any writes have occurred between the original time that these span
+// were read and the timestamp that the transaction now wants to commit at that
+// change the result of these reads. If any read would produce a different
+// result at the newer commit timestamp, the refresh fails and the transaction
+// is forced to fall back to a full transaction restart. However, if all of the
+// reads would produce exactly the same result at the newer commit timestamp,
+// the timestamp cache entries for these reads are updated and the transaction
+// is free to update its provisional commit timestamp without needing to
+// restart.
+//
+// TODO(nvanbenschoten): Unit test this file.
 type txnSpanRefresher struct {
 	st      *cluster.Settings
 	knobs   *ClientTestingKnobs

